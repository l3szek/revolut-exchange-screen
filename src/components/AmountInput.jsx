// @flow
import React from 'react';
import { focusOnContentEnd, getFormattedValue } from '../utils/helpers';
import { KEYCODES } from '../constants/common';
import TextField from '@material-ui/core/TextField';

import { formattedAmount } from '../utils/helpers';


type Props = {
  onChangeAmount: Function,
  currentValue: string | number,
  onFocus?: Function,
  focus?: boolean,
  maxWholePartLength?: number,
  maxDecimalPartLength?: number,
  name?: string,
  id?: string,
  classes?: string,
  pattern?: string,
  disabled?: boolean,
  maxLength?: string,
  autoFocus?: boolean,
}

class AmountInput extends React.Component<Props> {
  constructor(props) {
    super(props);
    const initialValue = props.currentValue > 0 ? props.currentValue : '';
    this.state = {
      value: this.getStateWithFormattedValue(initialValue),
      focused: false
    };
  }

  componentDidUpdate(prevProps) {
    if (prevProps.currentValue !== this.props.currentValue) {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({
        value: this.getStateWithFormattedValue(this.props.currentValue)
      });
    }
    if (this.props.focus) {
      focusOnContentEnd(this.input);
    }
  }

  /*
  static getDerivedStateFromProps(props, state) {
    if (props.currentValue !== state.value) {
      return {
        value: getFormattedValue(props.currentValue, 10, 20)
      };
    }
    return null;
  }

  */

  onKeydown = (event) => {
    const target = event.currentTarget;
    if (event.keyCode === KEYCODES.ENTER) {
      target.blur();
    }
  };

  onChange = (event) => {
    const newState = this.getStateWithFormattedValue(event.target.value);

    if (newState!== this.state.value) {
      // user allowed to type the point on the right side, but the
      // component does not count it when giving result number outside
      const eventValue = newState.endsWith('.') ? newState.slice(0, -1) : newState;

      event.target.value = eventValue;
      this.props.onChangeAmount(event);
    }

    this.setState({
      value: newState
    });
  }

  onFocus = () => {
    this.setState({
      focused: !this.state.focused
    });
  }

  getStateWithFormattedValue = (currentValue) => getFormattedValue(currentValue)
  

  input: ?HTMLDivElement;

  render() {
    const { classes, disabled, id, maxLength, name, pattern, autoFocus } = this.props;
    const { value, focused } = this.state;
    const amountFormatted = formattedAmount(value);

    const attributes = {
      autoCapitalize: 'off',
      autoComplete: 'off',
      autoCorrect: 'off',
      className: classes,
      disabled,
      id,
      maxLength,
      name,
      onKeyDown: this.onKeydown,
      onChange: this.onChange,
      onFocus: this.onFocus,
      onBlur: this.onFocus,
      pattern,
      ref: (e) => { this.input = e; },
      spellCheck: 'false',
      type: 'text',
      value: focused ? value : amountFormatted,
    };

    return (
      <TextField
        id="s-bssare"
        className={classes}
        margin="normal"
        inputProps={{ 'aria-label': 'amount-input' }}
        {...attributes}
        autoFocus={autoFocus}
      />
    )
  }
}

AmountInput.defaultProps = {
  currentValue: '',
  onFocus: null,
  focus: false,
  name: '',
  id: '',
  classes: '',
  pattern: '[0-9.]*',
  disabled: false,
};

export default AmountInput;
