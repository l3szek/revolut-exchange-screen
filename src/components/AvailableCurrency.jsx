// @flow

import React, {Fragment} from 'react';
import Typography from '@material-ui/core/Typography';

import { formattedAmount } from '../utils/helpers';

type Props = {
  symbol: string,
  availableMoney: number,
}

const AvailableCurrency = (props: Props) => { 
  const { symbol, availableMoney } = props;
  const amount = formattedAmount(availableMoney);
  return (
    <Fragment>
      <Typography
        variant="subtitle2"
        color="textSecondary"
      >
        Balance:
      </Typography>
      <Typography
        variant="subtitle1"
        color="textSecondary"
      >
        {symbol} {amount}
      </Typography>
    </Fragment>
  )
}

export default AvailableCurrency;