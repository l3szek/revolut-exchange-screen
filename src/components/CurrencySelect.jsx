// @flow

import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import MenuItem from '@material-ui/core/MenuItem';
import TextField from '@material-ui/core/TextField';

type Props = {
  currencies: Array<any>,
  classeNames: string,
  onCurrencyChange: SyntheticInputEvent<*> => any,
  currency: string,
}

const useStyles = makeStyles(theme => ({
  menu: {
    width: 150,
  },
}));

const CurrencySelect = (props: Props) => { 
  const { currencies, classeNames, onCurrencyChange, currency } = props;
  const classes = useStyles();
  const mappedCurrencies =
    currencies.map(option => (
      <MenuItem key={option.currency} value={option.currency}>
        {option.currency}
      </MenuItem>
    ));
  return (
    <TextField
      id="standard-select-currency"
      select
      className={`${classeNames} ${classes.menu}`}
      value={currency}
      onChange={(e) => onCurrencyChange(e)}
      SelectProps={{
        MenuProps: {
          className: classes.menu,
        },
      }}
      margin="normal"
    >
      {mappedCurrencies}
    </TextField>
  )
}

export default CurrencySelect;