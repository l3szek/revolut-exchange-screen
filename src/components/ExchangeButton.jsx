// @flow

import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Grid from '@material-ui/core/Grid';
import Fab from '@material-ui/core/Fab';

const useStyles = makeStyles(theme => ({
  grid: {
    marginTop: theme.spacing(2.5),
  },
  fab: {
    margin: theme.spacing(1),
    paddingLeft: '15%',
    paddingRight: '15%',
  },
}));

type Props = {
  onClick: Function,
  disabled: boolean,
}

const ExchangeButton = (props: Props) => { 
  const { onClick, disabled } = props;
  const classes = useStyles();
  return (
    <Grid
      container
      justify="center"
      className={classes.grid}
    >
      <Fab 
        variant="extended"
        color="secondary"
        aria-label="exchange"
        className={classes.fab}
        onClick={onClick}
        disabled={disabled}
      >
        Exchange
      </Fab >
    </Grid>
  )
}

export default ExchangeButton;