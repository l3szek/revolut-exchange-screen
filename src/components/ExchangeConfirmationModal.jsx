import React from 'react';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import DialogActions from '@material-ui/core/DialogActions';
import DialogContent from '@material-ui/core/DialogContent';
import DialogContentText from '@material-ui/core/DialogContentText';
import DialogTitle from '@material-ui/core/DialogTitle';
import { formattedAmount } from '../utils/helpers';

type Props = {
  open: Function,
  handleClose: Function,
  amountFrom: string,
  amountTo: string,
  currencyFrom: string,
  currencyTo: string,
}

const ExchangeConfirmationModal = (props: Props) => {
  const { open, handleClose, amountFrom, currencyFrom, amountTo, currencyTo } = props;
  const formattedAmountFrom = formattedAmount(amountFrom, 2);
  const formattedAmountTo = formattedAmount(amountTo, 2);
  return (
    <Dialog
      open={open}
      onClose={handleClose}
      aria-labelledby="alert-dialog-title"
      aria-describedby="alert-dialog-description"
    >
      <DialogTitle color="primary" id="alert-dialog-title">{"Exchange successful"}</DialogTitle>
      <DialogContent>
        <DialogContentText id="alert-dialog-description">
          You have successfully exchanged {currencyFrom}<strong>{formattedAmountFrom}</strong> to {currencyTo}<strong>{formattedAmountTo}</strong>
        </DialogContentText>
      </DialogContent>
      <DialogActions>
        <Button onClick={handleClose} color="primary">
          Thanks!
        </Button>
      </DialogActions>
    </Dialog>
  );
}

export default ExchangeConfirmationModal;