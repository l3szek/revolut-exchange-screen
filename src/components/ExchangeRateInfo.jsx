// @flow

import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Icon from '@material-ui/core/Icon';
import Typography from '@material-ui/core/Typography';
import Box from '@material-ui/core/Box';
import { formattedAmount } from '../utils/helpers';

import { theme as mainTheme } from '../utils/theme';

type Props = {
  symbol: string,
  rate: number| null,
  currencyTo: string,
}

const useStyles = makeStyles(theme => ({
  rate: {
    display: 'flex',
    border: `2px solid ${mainTheme.colors.grey}`,
    borderRadius: 25,
    width: 'auto',
    padding: '7px 10px 4px 10px',
    zIndex: 100,
    position: 'absolute',
    bottom: -20,
    left: '50%',
    marginLeft: -82,
    background: theme.palette.common.white
  },
}));

const ExchangeRateInfo = (props: Props) => { 
  const { symbol, rate, currencyTo } = props;
  const classes = useStyles();
  const amountFormatted = formattedAmount(rate);
  return (
    <Box className={classes.rate}>
      <Icon color="primary">
        show_chart
      </Icon>
      <Typography color="primary" variant="body2">
        1 {symbol} = {amountFormatted} {currencyTo}
      </Typography>
    </Box>
  )
}

export default ExchangeRateInfo;