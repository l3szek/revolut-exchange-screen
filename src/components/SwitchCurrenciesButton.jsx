// @flow

import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Icon from '@material-ui/core/Icon';

import { theme as mainTheme } from '../utils/theme';

type Props = {
  onCurrencySwitch: Function,
}

const useStyles = makeStyles(theme => ({
  switchBtn: {
    display: 'block',
    border: `2px solid ${mainTheme.colors.grey}`,
    borderRadius: '50%',
    height: theme.spacing(5),
    width: theme.spacing(5),
    paddingLeft: theme.spacing(0.75),
    paddingTop: theme.spacing(0.75),
    zIndex: 100,
    cursor: 'pointer',
    position: 'absolute',
    bottom: -theme.spacing(2.5),
    background: theme.palette.common.white
  },
}));

const SwitchCurrenciesButton = (props: Props) => { 
  const { onCurrencySwitch } = props;
  const classes = useStyles();
  return (
    <button
      onClick={onCurrencySwitch}
      className={classes.switchBtn}
    >
      <Icon color="primary">
        import_export
      </Icon>
    </button>
  )
}

export default SwitchCurrenciesButton;