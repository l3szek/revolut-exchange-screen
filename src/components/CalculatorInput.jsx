// @flow

import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';

import AmountInput from './AmountInput';
import CurrencySelect from './CurrencySelect';


type Props = {
  onChangeAmount: Function,
  currentValue: string | number,
}

const useStyles = makeStyles(theme => ({
  marginTop10: {
    marginTop: 20,
  },
  root: {
    flexGrow: 1,
  },
  paper: {
    height: 140,
    width: 100,
  },
  control: {
    padding: theme.spacing(2),
  },
  textField: {
    '& .MuiInput-formControl': {
      fontSize: 40,
    }
  },
  width: {
    width: 300,
  }
}));

const CalculatorInput = (props: Props) => { 
  const { onChangeAmount, currentValue, currencies } = props;
  const classes = useStyles();
  return (
    <Grid container justify="space-between" spacing={1} style={{ backgroundColor: '#cfe8fc', height: '100vh' }}>
      <Grid item>
        <CurrencySelect
          currencies={currencies}
          classeNames={classes.textField}
        />
        <Grid classes={classes.marginTop10} item>
          <Typography
            variant="subtitle2"
            color="textSecondary">
            Available currency:
          </Typography>
          <Typography variant="subtitle1" color="textSecondary">
            $233.24
          </Typography>
        </Grid>
      </Grid>
      <Grid item>
        <AmountInput
          onChangeAmount={onChangeAmount}
          currentValue={currentValue}
          classes={`${classes.textField} ${classes.width}`}
        />
        
      </Grid>
    </Grid> 
  )
}

export default CalculatorInput;