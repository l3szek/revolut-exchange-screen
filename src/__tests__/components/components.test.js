import React from 'react';
import renderer from 'react-test-renderer';

import { CURRENCY } from '../../constants/common';
import wallet from '../../utils/userWalletMock.json';

import AmountInput from '../../components/AmountInput';
import AvailableCurrency from '../../components/AvailableCurrency';
import CalculatorInput from '../../components/CalculatorInput';
import CurrencySelect from '../../components/CurrencySelect';
import ExchangeButton from '../../components/ExchangeButton';
import ExchangeConfirmationModal from '../../components/ExchangeConfirmationModal';
import ExchangeRateInfo from '../../components/ExchangeRateInfo';
import SwitchCurrenciesButton from '../../components/SwitchCurrenciesButton';

test('Create components snapshots', () => {
  const AmountInputComponent = renderer.create(
    <AmountInput currentValue="1234.50" />,
  );
  const AvailableCurrencyComponent = renderer.create(
    <AvailableCurrency
      symbol="$"
      availableMoney="1234.50" />,
  );
  const CalculatorInputComponent = renderer.create(
    <CalculatorInput
      userWallet={wallet.userWallet}
      currency={CURRENCY.USD}
      selectedCurrency={{
        availableMoney: 1000,
        symbol: "$"
      }}
    />,
  );
  const CurrencySelectComponent = renderer.create(
    <CurrencySelect
      currency={CURRENCY.EUR}
      currencies={wallet.userWallet} />,
  );
  const ExchangeButtonComponent = renderer.create(
    <ExchangeButton />,
  );
  const ExchangeConfirmationModalComponent = renderer.create(
    <ExchangeConfirmationModal
      amountFrom="100"
      currencyFrom={CURRENCY.EUR}
      amountTo="404.50"
      currencyTo={CURRENCY.PLN}
      open={false}
    />,
  );
  const ExchangeRateInfoComponent = renderer.create(
    <ExchangeRateInfo
      symbol="$"
      rate="1"
      currencyTo={CURRENCY.PLN}
    />,
  );
  const SwitchCurrenciesButtonComponent = renderer.create(
    <SwitchCurrenciesButton />,
  );

  let AmountInputComponentTree = AmountInputComponent.toJSON();
  let AvailableCurrencyComponentTree = AvailableCurrencyComponent.toJSON();
  let CalculatorInputComponentTree = CalculatorInputComponent.toJSON();
  let CurrencySelectComponentTree = CurrencySelectComponent.toJSON();
  let ExchangeButtonComponentTree = ExchangeButtonComponent.toJSON();
  let ExchangeConfirmationModalComponentTree = ExchangeConfirmationModalComponent.toJSON();
  let ExchangeRateInfoComponentTree = ExchangeRateInfoComponent.toJSON();
  let SwitchCurrenciesButtonComponentTree = SwitchCurrenciesButtonComponent.toJSON();

  expect(AmountInputComponentTree).toMatchSnapshot();
  expect(AvailableCurrencyComponentTree).toMatchSnapshot();
  expect(CalculatorInputComponentTree).toMatchSnapshot();
  expect(CurrencySelectComponentTree).toMatchSnapshot();
  expect(ExchangeButtonComponentTree).toMatchSnapshot();
  expect(ExchangeConfirmationModalComponentTree).toMatchSnapshot();
  expect(ExchangeRateInfoComponentTree).toMatchSnapshot();
  expect(SwitchCurrenciesButtonComponentTree).toMatchSnapshot();

});