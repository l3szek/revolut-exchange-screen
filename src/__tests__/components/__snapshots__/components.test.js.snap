// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create components snapshots 1`] = `
<div
  autoCapitalize="off"
  autoCorrect="off"
  className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal"
  onKeyDown={[Function]}
  pattern="[0-9.]*"
  spellCheck="false"
>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      aria-label="amount-input"
      autoComplete="off"
      className="MuiInputBase-input MuiInput-input"
      disabled={false}
      id="s-bssare"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      type="text"
      value="1,234.5"
    />
  </div>
</div>
`;

exports[`Create components snapshots 2`] = `
<h6
  className="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorTextSecondary"
>
  Balance: 
  $
   
  1,234.5
</h6>
`;

exports[`Create components snapshots 3`] = `
<div
  className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 MuiGrid-justify-xs-space-between"
>
  <div
    className="MuiGrid-root MuiGrid-item"
  >
    <div
      className="MuiFormControl-root MuiTextField-root makeStyles-textField-72 makeStyles-menu-177 MuiFormControl-marginNormal"
    >
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <div
          aria-haspopup="listbox"
          className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input MuiInputBase-inputSelect"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          role="button"
          tabIndex={0}
        >
          USD
        </div>
        <input
          id="standard-select-currency"
          onAnimationStart={[Function]}
          type="hidden"
          value="USD"
        />
        <svg
          aria-hidden="true"
          className="MuiSvgIcon-root MuiSelect-icon"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item"
    >
      <h6
        className="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorTextSecondary"
      >
        Balance: 
        $
         
        1,000
      </h6>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-item"
  >
    <div
      autoCapitalize="off"
      autoCorrect="off"
      className="MuiFormControl-root MuiTextField-root makeStyles-textField-72 makeStyles-amountInput-73 MuiFormControl-marginNormal"
      onKeyDown={[Function]}
      pattern="[0-9.]*"
      spellCheck="false"
    >
      <div
        className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
        onClick={[Function]}
      >
        <input
          aria-invalid={false}
          aria-label="amount-input"
          autoComplete="off"
          className="MuiInputBase-input MuiInput-input"
          disabled={false}
          id="s-bssare"
          onAnimationStart={[Function]}
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          required={false}
          type="text"
          value="0"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Create components snapshots 4`] = `
<div
  className="MuiFormControl-root MuiTextField-root undefined makeStyles-menu-177 MuiFormControl-marginNormal"
>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
    onClick={[Function]}
  >
    <div
      aria-haspopup="listbox"
      className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input MuiInputBase-inputSelect"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={0}
    >
      EUR
    </div>
    <input
      id="standard-select-currency"
      onAnimationStart={[Function]}
      type="hidden"
      value="EUR"
    />
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root MuiSelect-icon"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
  </div>
</div>
`;

exports[`Create components snapshots 5`] = `
<div
  className="MuiGrid-root makeStyles-grid-199 MuiGrid-container MuiGrid-justify-xs-center"
>
  <button
    aria-label="exchange"
    className="MuiButtonBase-root MuiFab-root makeStyles-fab-200 MuiFab-extended MuiFab-secondary"
    disabled={false}
    onBlur={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiFab-label"
    >
      Exchange
    </span>
    <span
      className="MuiTouchRipple-root"
    />
  </button>
</div>
`;

exports[`Create components snapshots 6`] = `null`;

exports[`Create components snapshots 7`] = `
<div
  className="MuiBox-root MuiBox-root-241 makeStyles-rate-239"
>
  <span
    aria-hidden={true}
    className="material-icons MuiIcon-root MuiIcon-colorPrimary"
  >
    show_chart
  </span>
  <p
    className="MuiTypography-root MuiTypography-body2 MuiTypography-colorPrimary"
  >
    1 
    $
     = 
    1
     
    PLN
  </p>
</div>
`;

exports[`Create components snapshots 8`] = `
<button
  className="makeStyles-switchBtn-251"
>
  <span
    aria-hidden={true}
    className="material-icons MuiIcon-root MuiIcon-colorPrimary"
  >
    import_export
  </span>
</button>
`;
